base/MAPL_Resource.F90

```diff
@@ -0,0 +1,411 @@
+
+   ! MAPL searches for labels with certain prefixes as well as just the label itself
+   pure function get_labels_with_prefix(compname, label) result(labels_with_prefix)
+      character(len=*), intent(in) :: compname, label
+      character(len=ESMF_MAXSTR) :: labels_with_prefix(4), component_type
+
+      component_type = compname(index(compname, ":") + 1:)
+
+      ! The order to search for labels in resource files
+      labels_with_prefix(1) = trim(compname)//"_"//trim(label)
+      labels_with_prefix(2) = trim(component_type)//"_"//trim(label)
+      labels_with_prefix(3) = trim(label)
+      labels_with_prefix(4) = trim(compname)//MAPL_CF_COMPONENT_SEPARATOR//trim(label)
+
+   end function get_labels_with_prefix
+
+   ! If possible, find label or label with prefix. Out: logical to if label found, 
+   ! version of label found,
+   subroutine get_label_to_use(config, label, compname, label_is_present, label_to_use, rc)
+      type(ESMF_Config), intent(inout) :: config
+      character(len=*), intent(in) :: label
+      character(len=*), intent(in) :: compname
+      logical, intent(out) :: label_is_present
+      character(len=:), allocatable, intent(out) :: label_to_use
+      integer, optional, intent(out) :: rc
+
+      character(len=ESMF_MAXSTR), allocatable :: labels_to_try(:)
+      integer :: i
+      integer :: status
+
+      label_is_present = .false.
+      labels_to_try = get_labels_with_prefix(compname, label)
+
+      do i = 1, size(labels_to_try)
+         label_to_use = trim(labels_to_try(i))
+         call ESMF_ConfigFindLabel(config, label = label_to_use, isPresent = label_is_present, _RC)
+
+         if (label_is_present) then
+            exit
+         end if
+      end do
+
+      _RETURN(_SUCCESS)
+
+   end subroutine get_label_to_use
+
+   ! Find value of scalar variable in config
+   subroutine MAPL_GetResource_config_scalar(config, val, label_to_find, default, compname, rc)
+      type(ESMF_Config), intent(inout) :: config
+      class(*), intent(inout) :: val
+      character(len=*), intent(in) :: label_to_find
+      class(*), optional, intent(in) :: default
+      character(len=*), optional, intent(in) :: compname
+      integer, optional, intent(out) :: rc
+
+      integer :: status, printrc
+      logical :: default_is_present, label_is_present
+      character(len=:), allocatable :: label_to_print
+      character(len=:), allocatable :: label
+
+      default_is_present = present(default)
+
+      if (default_is_present) then
+         _ASSERT(same_type_as(val, default), "Value and default must have same type")
+      end if
+
+      ! If compname is present, find label in some form in config. Else search
+      ! for exact label
+      if (present(compname)) then
+         call get_label_to_use(config, label_to_find, compname, label_is_present, label, _RC)
+      else
+         label = label_to_find
+         call ESMF_ConfigFindLabel(config, label = label, isPresent = label_is_present, _RC)
+      end if
```

Comment:
I think that `get_label_to_use` should make the `compname` argument optional.   This will simplify the higher level logic here by eliminating the conditional block.
```suggestion
      call get_label_to_use(config, label_to_find, compname, label_is_present, label, _RC)
```
You will unfortunately need to reorder the arguments as well, as I would prefer that the optional arguments be after the non-optional.  Not a public interface, se we have that flexibility.   Indeed, probably should throw in a `KeywordEnforcer` argument for good form.

Reply:
Done